### ISSUE TITLE
Potentially corrupt connection in pool

### ISSUE THREAD

**Issue by jebeaudet** [awslabs/aws-mysql-jdbc#480]
### Describe the bug

We have a global aurora cluster with the following topology : 
us-east-1 : 
- 1 writer
- 1 reader

us-east-2 + eu-west-1 + ap-southeast-2 + ca-central-1 : 
- 2 readers in each region

During the latest round of OS patches in our global aurora cluster, we had to failover the writer node in order to apply them. During that failover, it somehow triggered a reboot of the 2 reader instances in the us-east-2 region. 

The service instances in us-east-2 use the local reader endpoint to direct RO transactions. After the reboot, we had 2 of the 4 service instances that were constantly throwing errors and not auto healing. We had to reboot them to get things back on track. They all got the `08S02` error code which is configured to not evict the connection since the underlying driver one has successfully failover to another node.

I've inspected the code a bit and unfortunately found nothing weird in the concurrency process of the failover. However, I noticed that all the stack trace were identical, everything was failing in the `setNetworkTimeout` method. Do you know if there are any limitations or anything within the driver that would produce such result?

### Expected Behavior

The service's connections should auto heal.

### Current Behavior

The service was constantly handling exceptions with the `08S02` error code and producing errors.

### Reproduction Steps

I unfortunately can't reproduce.

### Additional Information/Context

Here is the tidied stack trace : 
```
Caused by: java.sql.SQLTransientConnectionException: Secondary-HikariPool - Connection is not available, request timed out after 5000ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:197)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:162)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
	at org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource.getConnection(AbstractRoutingDataSource.java:194)
	at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)
	at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:38)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:108)
	... 218 common frames omitted
Caused by: java.sql.SQLException: The active SQL connection has changed due to a connection failure. Please re-configure session state if required.
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.plugins.failover.FailoverConnectionPlugin.failover(FailoverConnectionPlugin.java:509)
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.plugins.failover.FailoverConnectionPlugin.pickNewConnection(FailoverConnectionPlugin.java:652)
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.plugins.failover.FailoverConnectionPlugin.pickNewConnection(FailoverConnectionPlugin.java:634)
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.plugins.failover.FailoverConnectionPlugin.dealWithOriginalException(FailoverConnectionPlugin.java:875)
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.plugins.failover.FailoverConnectionPlugin.execute(FailoverConnectionPlugin.java:283)
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.plugins.ConnectionPluginManager.execute(ConnectionPluginManager.java:138)
	at software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ha.ConnectionProxy.invoke(ConnectionProxy.java:196)
	at jdk.proxy3/jdk.proxy3.$Proxy176.setNetworkTimeout(Unknown Source)
	at com.zaxxer.hikari.pool.PoolBase.setNetworkTimeout(PoolBase.java:566)
	at com.zaxxer.hikari.pool.PoolBase.isConnectionAlive(PoolBase.java:173)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:186)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:162)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)```

### The AWS JDBC Driver for MySQL version used

1.1.10

### JDK version used

openjdk 17.0.8.1 2023-08-24 LTS OpenJDK Runtime Environment Zulu17.44+53-CA (build 17.0.8.1+1-LTS) OpenJDK 64-Bit Server VM Zulu17.44+53-CA (build 17.0.8.1+1-LTS, mixed mode, sharing)

### Operating System and version

ubuntu 22.04
```[tasklist]
### Tasks
```

**Comment by crystall-bitquill** [comment]
Hi @jebeaudet,

Thanks for reaching out and raising this issue.

To clarify, when you triggered the failover process, was it for the primary cluster or for the global database? The AWS JDBC Driver for MySQL does not support global databases, and [new features will not be added going forward](https://github.com/awslabs/aws-mysql-jdbc/blob/main/RELEASE_POLICY.md). Triggering failover for the global database will result in errors.

Although the driver does not support global databases, it is still possible to perform a failover on the primary cluster. To test this, I set up a global database with the primary cluster containing one writer and one reader, and two secondary clusters in separate regions with two readers each (this was using the engine version Aurora MySQL 3.04.0). I triggered failover for the primary cluster and the failover process was completed successfully. I had three connections pools, one to the primary cluster and one to the reader endpoint for each secondary cluster. I have not experienced any rebooting in the secondary clusters. Is there anything you think may be missing or different from this setup, besides having fewer secondary clusters?

You also mentioned one of your secondary clusters started rebooting after triggering failover, and that two of the four service instances were constantly throwing errors. Were the errors from the secondary cluster that had been rebooted?

It's difficult to determine the root cause, but one suggestion would be to increase your `connectionTimeout` value in your HikariCP configuration. Since the error message you got seems to indicate failover was successful but the connection timed out, it's possible that when a new connection is retrieved from the connection pool after the writer failover, the reader failover process takes long enough that the connection times out.

**Comment by jebeaudet** [comment]
Hi @crystall-bitquill !

Thanks for the `aws-advanced-jdbc-wrapper` info, I missed that from the release policy. I strongly suggest you mention it in the readme of this repository, there is absolutely no mention of the july maintenance date and the feature freeze. Release notes would also be another good place to mention it.

As for this issue, yes it was simply a failover of the primary cluster. It was an aurora 2.11.1 cluster. From what we gathered, the reboot process of secondary cluster seems normal (you have to check the mysql logs to see it, it will not show in the aws console or events) so I think we can ignore that from the problem.

As for the errors, they only happened on 2 of the 4 service instances in us-east-2 with the primary datasource. Secondary datasource in us-east-2 (and all other regions) auto healed normally. 

Since the connections were in a broken state for very long (long after the failover completed), do you really thing the connection timeout could be in play here? It really seems like an internal issue with the internal connection state.

Thanks a lot for checking this out, I'm fully aware this is not the easiest problem and I opened the issue the in hope that you might find a corner case in the code that I didn't.

**Comment by crystall-bitquill** [comment]
Hi @jebeaudet,

Thank you for the documentation feedback! I've merged some changes (PR #482) to make that information more visible to users.

I'm afraid I also have not found anything in the failover code that might lead to these errors.

I suggested the connection timeout as a potential problem because of the exceptions in the stack trace. If the failover exception with the code `08S02` was thrown, then a successful connection should have been created. If there weren't any other errors between getting the successful failover results and getting the timeout error, it seemed like a possible cause. 

If it were an internal connection state issue, that would mean either the failover results were incorrect and the connection was not actually valid, or that after obtaining the failover results something else happened that led to `setNetworkTimeout` timing out. I haven't found anything that would indicate the first case can occur, and I can't tell at the moment what might lead to the second case.

Lastly, I do have a few follow up questions:
1. To clarify, 2 out of 4 service instances had issues, but you mentioned only having two readers in the us-east-2 region? 
2. How were the datasources configured?
3. How long were the connections in a broken state? 
4. Would you happen to have any driver logs saved from the event?
